tests/Model/LazyBSONArrayTest.php

```diff
@@ -0,0 +1,289 @@
+    {
+        $this->assertInstanceOf(LazyBSONDocument::class, $array[1]);
+        $this->assertInstanceOf(LazyBSONDocument::class, $array[1]);
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testGetArray(LazyBSONArray $array): void
+    {
+        $this->assertInstanceOf(LazyBSONArray::class, $array[2]);
+        $this->assertInstanceOf(LazyBSONArray::class, $array[2]);
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetExists(LazyBSONArray $array): void
+    {
+        $this->assertTrue(isset($array[0]));
+        $this->assertFalse(isset($array[4]));
+
+        // Unsupported offset
+        $this->assertFalse(isset($array['foo']));
+
+        // Numeric offset
+        $this->assertTrue(isset($array['1']));
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetSet(LazyBSONArray $array): void
+    {
+        $this->assertFalse(isset($array[4]));
+        $array[4] = 'yay!';
+        $this->assertSame('yay!', $array[4]);
+
+        $this->assertSame('bar', $array[0]);
+        $array[0] = 'baz';
+        $this->assertSame('baz', $array[0]);
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetSetForNumericOffset(LazyBSONArray $array): void
+    {
+        $array['1'] = 'baz';
+        $this->assertSame('baz', $array[1]);
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetSetForUnsupportedOffset(LazyBSONArray $array): void
+    {
+        $this->expectWarning();
+        $this->expectWarningMessage('Unsupported offset: foo');
+        $array['foo'] = 'yay!';
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testAppend(LazyBSONArray $array): void
+    {
+        $this->assertFalse(isset($array[3]));
+        $array[] = 'yay!';
+        $this->assertSame('yay!', $array[3]);
+    }
+
+    public function testAppendToEmptyArray(): void
+    {
+        $array = new LazyBSONArray();
+
+        $this->assertFalse(isset($array[0]));
+        $array[] = 'yay!';
+        $this->assertSame('yay!', $array[0]);
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testAppendWithGap(LazyBSONArray $array): void
+    {
+        // Leave offset 3 empty
+        $array[4] = 'yay!';
+
+        $this->assertFalse(isset($array[3]));
+        $array[] = 'bleh';
+
+        // Expect offset 3 to be skipped, offset 5 is used as 4 is already set
+        $this->assertFalse(isset($array[3]));
+        $this->assertSame('bleh', $array[5]);
+    }
+
+    /** @dataProvider provideTestArray */
+    public function testOffsetUnset(LazyBSONArray $array): void
+    {
+        $this->assertFalse(isset($array[4]));
+        $array[4] = 'yay!';
+        unset($array[4]);
+        $this->assertFalse(isset($array[4]));
+
+        unset($array[0]);
+        $this->assertFalse(isset($array[0]));
+
+        // Change value to ensure it is unset for good
```

Comment:
What does "unset for good" mean in this context? You're doing the same assertion here as above.

I understand this is going to assign something to the `$set` property and later clear it. The `$read` property (if it was ever set) will remain as-is.

Reply:
I wanted to specifically test that pre-existing values are considered unset even when they've been changed before. I've updated the comment to convey this here and in the corresponding document test.